README.md
<p align="center">
  <img src="assets/logo.png" alt="pino to kafka"/>
</p>

# pino-kafka

This module provides a "transport" for [pino][pino] that simply forwards
messages to kafka.

You should install `pino-kafka` globally for ease of use:

```shell
$ npm install --production -g pino-kafka
### or with yarn
$ yarn global add pino-kafka
```

[pino]: https://www.npmjs.com/package/pino

## Requirements
This library depends on `node-rdkafka`.
Have a look at [node-rdkafka requirements](https://github.com/Blizzard/node-rdkafka#requirements).


## Usage
### CLI
Given an application `foo` that logs via [pino][pino], and a kafka broker listening on `10.10.10.5:9200` you would use `pino-kafka` as:

```bash
$ node foo | pino-kafka -b 10.10.10.5:9200
```

### Programmatic Usage
Initialize `pino-kafka` and pass it to `pino`.
```js
const pino = require('pino')
const pkafka = require('pino-kafka')

const logger = pino({}, pkafka({ brokers: "10.10.10.5:9200"}))
```
## Options
+ `--brokers` (`-b`): broker list for kafka producer. Comma separated
+ `--defaultTopic` (`-d`): default topic name for kafka.
+ `--timeout` (`-t`): timeout for initial broker connection in milliseconds. Default 10000
+ `--echo` (`-e`): echo the received messages to stdout. Default: false.
+ `--settings`: path to config JSON file. Have a look at [Settings JSON file](#settings-json-file) section for details and examples
+ `--kafka.$config`: any kafka configuration can be passed with prefix `kafka`. Please visit [node-rdkafka configuration](https://github.com/edenhill/librdkafka/blob/v1.3.0/CONFIGURATION.md) for available options.
Note that only producer and global configuration properties will be used.
Have a look at [Kafka Settings](#kafka-settings) section for details and examples


### Settings JSON File

The `--settings` switch can be used to specify a JSON file that contains
a hash of settings for the application. A full settings file is:

```json
{
  "brokers": "10.6.25.11:9092, 10.6.25.12:9092",
  "defaultTopic": "blackbox",
  "kafka": {
    "compression.codec":"none",
    "enable.idempotence": "true",
    "max.in.flight.requests.per.connection": 4,
    "message.send.max.retries": 10000000,
    "acks": "all"
  }
}
```

Note that command line switches take precedence over settings in a settings
file. For example, given the settings file:

```json
{
  "brokers": "my.broker",
  "defaultTopic": "test"
}
```

And the command line:

```bash
$ yes | pino-kafka -s ./settings.json -b 10.10.10.11:9200
```

The connection will be made to address `10.10.10.11:9200` with the default topic `test`.

### Kafka Settings

You can pass `node-rdkafka` producer configuration by prefixing the property with `kafka.` For example:
```bash
$ yes | pino-kafka --kafka.retries=5 --kafka.retry.backoff.ms=500
```

In the Setting JSON File you can use followings:
```json
{
  "kafka": {
    "retries": "5",
    "retry.backoff.ms": "500"
  }
}
```

Following will work also:
```json
{
  "kafka": {
    "retries": "5",
    "retry":{
      "backoff": {
        "ms":  "500"
      }
    }
  }
}
```

## Accessing Internal Kafka Producer
You can access `node-rdkafka` producer from pino stream with `_kafka`.

For example:
```js
const pino = require('pino')
const pkafka = require('pino-kafka')

const pKafkaStream = pkafka({ brokers: "10.10.10.5:9200"})
const logger = pino({}, pKafkaStream)

// From pino-kafka instance
pKafkaStream._kafka.getMetadata({}, (err, data)=> {
    //...
})

// From logger
logger[pino.symbols.streamSym]._kafka.getMetadata({}, (err, data)=> {
    //...
})
```

## Testing
For running tests make sure you installed dependencies with `npm install` or `yarn` and have a running kafka.
More easily, if you have docker and docker-compose installed, you can create one with following.

```bash
$ cd pino-kafka
$ docker-compose up -d
```

Look at [docker-compose file](docker-compose.yml) for more details.

After you all setup, just run test command with following:
```bash
$ npm run test
# or with yarn
$ yarn test
```

> **NOTE**: If you use your own kafka setup, you may need to change test configuration accordingly to your needs(ip, topic etc.)

## License
[MIT](LICENSE)


CHANGELOG.md
## [0.1.2](https://github.com/ayZagen/pino-kafka/compare/v0.1.1...v0.1.2) (2021-04-10)


### Features

* **cli:** flush kafka queue on process exit ([5793b40](https://github.com/ayZagen/pino-kafka/commit/5793b40c750aa701e76a3ca52a746ff89a4860a5))

## [0.1.1](https://github.com/ayZagen/pino-kafka/compare/v0.1.0...v0.1.1) (2021-04-10)


### Features

* expose kafka stream ([a1a0e2f](https://github.com/ayZagen/pino-kafka/commit/a1a0e2faf70ab8a2906307c4472b20c2a74afc14))

# 0.1.0 (2021-01-18)


### Bug Fixes

* consistent cli commands and help file ([79c67e5](https://github.com/ayZagen/pino-kafka/commit/79c67e5848dc7d99ccdf073295c043bf0b445f1e))
* read stream does not wait for kafka producer to connect ([40cf06b](https://github.com/ayZagen/pino-kafka/commit/40cf06b53d4b49d9ccaf00da40ff4b73b76d776a))


### Features

* add events and timeout option ([9a54303](https://github.com/ayZagen/pino-kafka/commit/9a543030d9e8be2207ed82b9f69198e57c2f6171))
* add missing cli dependency pump ([d7c1bf2](https://github.com/ayZagen/pino-kafka/commit/d7c1bf2bc84a8fda57bbeff24d575956ed642c66))
* echoing false by default ([c6b3417](https://github.com/ayZagen/pino-kafka/commit/c6b3417c5a9a4c4441daa735aa00eb3fb6829db1))
* make printing to stdout configurable ([b338a66](https://github.com/ayZagen/pino-kafka/commit/b338a661655b12cc6722bc6222b0bd0572b1944e))


### Reverts

* "ci: add node 6.x and 8.x versions" ([10ee5c1](https://github.com/ayZagen/pino-kafka/commit/10ee5c1697fad39e4480aff0afcf972c649e2f93))

# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 0.0.1 (2020-03-23)
* Initial commit


test/pkafka.test.js
'use strict'

const pino = require('pino')
const {KafkaConsumer, CODES} = require('node-rdkafka')
const pKafka = require('../pkafka')
const assert = require('assert')
const options = {
  topic: 'test',
  brokers: 'localhost:29092',
  consumerGroup: 'kafka',
  timeout: 2000
}
let consumer

function initConsumer(done) {

  consumer = new KafkaConsumer({
    "metadata.request.timeout.ms": options.timeout,
    'socket.timeout.ms': options.timeout,
    'group.id': options.consumerGroup,
    'metadata.broker.list': options.brokers,
    'enable.auto.commit': false
  }, {
    'auto.offset.reset': 'earliest' // consume from the start
  })

  consumer.connect({timeout: options.timeout}, (err) => {
    if (err) {
      done(err)
    }
  })
  consumer
    .on('ready', () => {
      consumer.subscribe([options.topic]);
      this.consumer = consumer
      this.consumer.consume();
      this.consumer.commit()
      done(null, consumer)
    })
    .on('error', (err) => {
      done(err)
    })
}

describe('api', function () {
  it('should expose kafka producer', function () {
    const logger = pino(pKafka({
      brokers: options.brokers,
      timeout: options.timeout,
      defaultTopic: options.topic
    }))
    assert(logger[pino.symbols.streamSym]._kafka)
  });
})
describe('simple produce', function () {

  before(function (done) {
    initConsumer.call(this,done)
  })

  it('should write successfully to kafka topic', function (done) {
    const logger = pino(pKafka({
      brokers: options.brokers,
      timeout: options.timeout,
      defaultTopic: options.topic
    }))

    const messages = []

    for (let i = 0; i < 500; i++) {
      messages.push({index: `field${i}`, time: new Date().getTime()})
    }


    const incomingMessages = []
    this.consumer.on("data", data => {
      incomingMessages.push(JSON.parse(data.value.toString('utf8')))
      if (incomingMessages.length === messages.length) {
        assert(incomingMessages.length === 500)
        done()
      }
    })

    messages.forEach(msg => {
      logger.info(msg)
    })
  })

  after(function (done) {
    if (this.consumer) {
      this.consumer.disconnect(() => {
        done()
      })
    } else {
      done()
    }
  })

})


pkafka.js
const stream = require('stream')
const Kafka = require('node-rdkafka')
const util = require("util");

module.exports = function pinoKafka(opts) {

  opts.on = Object.assign({
    error(err) {
      process.stderr.write(util.format.apply(this, err) + '\n')
    },
    ready() {
    },
    disconnected(){},
    'event'(){},
    'event.log'(){},
    'event.stats'(){},
    'event.throttle'(){},
    'delivery-report'(){},
  }, opts.on || {})
  const through = new stream.PassThrough()
  const inputStream = process.stdin
  through.pause()

  const kafkaStream = new Kafka.HighLevelProducer({
    ...opts.kafka,
    'metadata.broker.list': opts.brokers
  })

  kafkaStream.connect({
    timeout: opts.timeout
  }, (err) => {
    if (err) {
      opts.on.error.call(kafkaStream, err)
    }
  })

  kafkaStream.on('ready', (info, metadata) => {
    through.pipe(outputStream)
    through.resume()
    opts.on.ready.call(kafkaStream, info, metadata)
  })

  kafkaStream.on('disconnected', opts.on.disconnected.bind(kafkaStream))
  kafkaStream.on('event', opts.on.event.bind(kafkaStream))
  kafkaStream.on('event.log', opts.on['event.log'].bind(kafkaStream))
  kafkaStream.on('event.stats', opts.on['event.stats'].bind(kafkaStream))
  kafkaStream.on('event.error', opts.on.error.bind(kafkaStream))
  kafkaStream.on('event.throttle', opts.on['event.throttle'].bind(kafkaStream))
  kafkaStream.on('delivery-report', opts.on['delivery-report'].bind(kafkaStream))

  const outputStream = new stream.Writable({
    write(body, enc, cb) {
      // TODO: remove new line delimeters
      kafkaStream.produce(opts.defaultTopic,
        null,
        body,
        null,
        null,
        (err, offset) => {
          if (err) {
            opts.on.error.call(kafkaStream, err)
            cb(err)
          } else {
            if (opts.echo){
              process.stdout.write(body)
            }
            cb()
          }
        })
    }
  })

  inputStream.pipe(through)

  through._kafka = kafkaStream

  return through;
}



cli.js
#! /usr/bin/env node
'use strict'

const minimist = require('minimist')
const pump = require('pump')
const fs = require('fs')
const path = require('path')
const pinoKafka = require('./pkafka')
const util = require("util");

function keysToDotNotation(obj, current, final) {
  if(!final) {
    final = {}
  }
  for (var key in obj) {
    var value = obj[key];
    var newKey = (current ? current + "." + key : key);  // joined key with dot
    if (value && typeof value === "object") {
      keysToDotNotation(value, newKey, final);  // it's a nested object, so do it again
    } else {
      final[newKey] = value;  // it's not an object, so set the property
    }
  }
  return final
}

function start (opts) {
  if(opts.kafka){
    if(typeof opts.kafka !== 'object'){
      throw new Error('Kafka options must be an object')
    }
    opts.kafka = keysToDotNotation(opts.kafka)
  }
  if (opts.help) {
    console.log(fs.readFileSync(path.join(__dirname, './help.txt'), 'utf8'))
    return
  }

  if (opts.version) {
    console.log('pino-kafka', require('./package.json').version)
    return
  }


  if (opts.settings) {
    try {
      const loadedSettings = require(path.resolve(opts.settings))
      const settings = Object.assign(loadedSettings, argv)
      opts = Object.assign(opts, settings)
    } catch (e) {
      console.error('`settings` parameter specified but could not load file: %s', e.message)
      process.exit(1)
    }
  }
  const stream = pinoKafka(opts)

  pump(process.stdin, stream)

  function terminator( sig ) {
    if ( typeof sig === 'string' ) {
      flushKafkaQueue(function (err) {
        process.exit(err ? 1 : 0);
      });
    }
  }

  async function flushKafkaQueue( callback ){
    stream._kafka.flush(opts.timeout, err => {
      if(err){
        process.stderr.write(util.format.apply(this, err) + '\n')
      }
      callback(err)
    })
  }

  ['SIGHUP', 'SIGINT', 'SIGQUIT', 'SIGILL', 'SIGTRAP', 'SIGABRT',
    'SIGBUS', 'SIGFPE', 'SIGUSR1', 'SIGSEGV', 'SIGUSR2', 'SIGTERM'
  ].forEach(function (sig) {
    process.on(sig, function () {
      try {
        terminator(sig);
      } catch (e) {
        process.exit(1)
      }
    });
  });
}

start(minimist(process.argv.slice(2), {
  alias: {
    version: 'v',
    help: 'h',
    brokers: 'b',
    defaultTopic: 'd',
    settings: 's',
    echo: 'e',
    timeout: 't'
  },
  default: {
    kafka: {
      'compression.codec':'none',
      'enable.idempotence': 'true',
      'max.in.flight.requests.per.connection': 4,
      'message.send.max.retries': 10000000,
      'acks': 'all'
    },
    echo: false,
    timeout: 10000,
    brokers: '10.6.25.11:9092, 10.6.25.12:9092',
    defaultTopic: 'blackbox'
  }
}))
